File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import sys
import math
input = sys.stdin.readline


def solution(fees, records):
    answer = []
    default_time, default_fee, unit_time, unit_fee = fees
    car_number_dic = {}
    for record in records:
        time, car_number, in_out = record.split(" ")
        if in_out == "IN":
            if car_number in car_number_dic:
                car_number_dic[car_number][0] = time
            else:
                car_number_dic[car_number] = [time, 0]
        else:
            in_time = car_number_dic[car_number][0]
            period = parse_to_minutes(time) - parse_to_minutes(in_time)
            car_number_dic[car_number][0] = -1
            car_number_dic[car_number][1] += period

    for k in car_number_dic.keys():
        if car_number_dic[k][0] != -1:
            period = parse_to_minutes(
                "23:59") - parse_to_minutes(car_number_dic[k][0])
            car_number_dic[k][1] += period
        t = car_number_dic[k][1]
        if default_time > t:
            answer.append([k, default_fee])
        else:
            fee = default_fee + math.ceil((t-default_time)/unit_time)*unit_fee
            answer.append([k, fee])

    answer.sort(key=lambda x: int(x[0]))
    return [x[1] for x in answer]


def parse_to_minutes(time):
    a, b = time.split(":")
    return int(a)*60 + int(b)

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int[] solution(int[] fees, String[] records) {
        Map<String, Integer> carTime = new HashMap<>();
        Map<String, Integer> carFee = new HashMap<>();

        for (String record : records) {
            String[] split = record.split(" ");
            int time = getTime(split[0]);
            String num = split[1];
            boolean in = split[2].equals("IN");

            if (in) {
                carTime.put(num, time);
            } else {
                int totalTime = time - carTime.get(num);
                carFee.put(num, carFee.getOrDefault(num, 0) + totalTime);
                carTime.remove(num);
            }
        }

        int lastOutTime = getTime("23:59");
        for (Map.Entry<String, Integer> entry : carTime.entrySet()) {
            String num = entry.getKey();
            int inTime = entry.getValue();

            int time = lastOutTime - inTime;

            carFee.put(num, carFee.getOrDefault(num, 0) + time);
        }

        // 출력
        for (Map.Entry<String, Integer> entry : carFee.entrySet()) {
            System.out.println(entry.getKey() + " " + entry.getValue());
        }

        return carFee.entrySet().stream()
                .sorted((o1, o2) -> o1.getKey().compareTo(o2.getKey()))
                .map(entry -> getFee(fees, entry.getValue()))
                .mapToInt(v -> v)
                .toArray();
    }

    public int getTime(String time) {
        String[] split = time.split(":");
        int h = Integer.parseInt(split[0]);
        int m = Integer.parseInt(split[1]);

        return h * 60 + m;
    }

    public int getFee(int[] fees, int time) {
        if (time <= fees[0]) {
            return fees[1];
        }

        int fee = fees[1];
        time -= fees[0];
        fee += (int) (Math.ceil((double) time / fees[2]) * fees[3]);

        return fee;
    }
}